rllib/tests/test_node_failure.py

```diff
@@ -0,0 +1,226 @@
+
+        # One step with all nodes up, enough to satisfy resource requirements
+        ppo.step()
+
+        self.assertEqual(len(ppo.workers._worker_health_check()), 0)
+        self.assertEqual(len(ppo.workers._remote_workers), 6)
+
+        # Remove the first non-head node.
+        node_to_kill = get_other_nodes(self.cluster, exclude_head=True)[0]
+        self.cluster.remove_node(node_to_kill)
+
+        assert len(ppo.workers._worker_health_check()) == 2
+        self.assertEqual(len(ppo.workers._remote_workers), 6)
+
+        # node comes back immediately.
+        self.cluster.add_node(
+            redis_port=None,
+            num_redis_shards=None,
+            num_cpus=2,
+            num_gpus=0,
+            object_store_memory=object_store_memory,
+            redis_max_memory=redis_max_memory,
+            dashboard_host="0.0.0.0",
+        )
+
+        # One step with a node down, resource requirements not satisfied anymore
+        # step() should hang trying to restore the missing workers.
+        ppo.step()
+
+        # Workers should be back up, everything back to normal.
+        self.assertEqual(len(ppo.workers._worker_health_check()), 0)
+        self.assertEqual(len(ppo.workers._remote_workers), 6)
+
+    def test_wait_for_nodes_on_failure(self):
+        # We tolerate failing workers and pause training
+        config = (
+            PPOConfig()
+            .rollouts(
+                num_rollout_workers=6,
+                ignore_worker_failures=False,
+                recreate_failed_workers=True,
+            )
+            .training()
+        )
+        ppo = PPO(config=config, env="CartPole-v0")
+
+        # One step with all nodes up, enough to satisfy resource requirements
+        ppo.step()
+
+        # Remove the first non-head node.
+        node_to_kill = get_other_nodes(self.cluster, exclude_head=True)[0]
+        self.cluster.remove_node(node_to_kill)
+
+        # Check faulty worker indices
+        # All nodes have 2 CPUs and we require a driver thread + 5 rollout workers
+        self.assertEqual(len(ppo.workers._worker_health_check()), 2)
+        self.assertEqual(len(ppo.workers._remote_workers), 6)
+
+        def _step_target():
+            # Resource requirements satisfied after approx 30s
+            ppo.step()
+
+        time_before_step = time.time()
+
+        # kill one node after n seconds
+        t = threading.Thread(target=_step_target)
+        t.start()
+
+        # Wait 30 seconds until the missing node reappears
+        time.sleep(30)
+        self.cluster.add_node(
+            redis_port=None,
+            num_redis_shards=None,
+            num_cpus=2,
+            num_gpus=0,
+            object_store_memory=object_store_memory,
+            redis_max_memory=redis_max_memory,
+            dashboard_host="0.0.0.0",
+        )
+
+        t.join()
+
+        td = time.time() - time_before_step
+
+        assert 30 < td < 60, (
```

Comment:
self.assertGreater and self.assertLesser man.

Reply:
:(